nextstrain/cli/util.py

```diff
@@ -461,3 +471,70 @@ def glob_match(path: Union[str, Path], patterns: Union[str, Sequence[str]]) -> b
     Implemented with with :func:`wcmatch.glob.globmatch`.
     """
     return globmatch(path, patterns, flags = GLOBSTAR | BRACE | EXTGLOB | MATCHBASE | NEGATE)
+
+
+def user_app_data_dir() -> Path:
+    """
+    Determines the appropriate directory for our user-specific application
+    data for the current user.
+
+    Supports macOS- and Windows-specific paths and follows the XDG spec for
+    all other systems (Linux and other Unixes).
+    """
+    system = platform.system()
+
+    # Paths from appdirs <https://github.com/ActiveState/appdirs/blob/master/appdirs.py>
+    if system == "Darwin":
+        user_data = Path.home() / "Library/Application Support"
```

Comment:
Nice spotting! I'll try to reproduce this locally. From the logs above, there appears to be a path handling bug in two post-install scripts associated with specific packages (possibly librsvg and graphviz, but not clear if output is mixed up a bit).

Reply:
Looks like also issues with the openmpi package:

```
Linking openmpi-4.1.2-hbfc84c5_0
/tmp/mambafhtXDHueer5: line 2: /tmp/hello: No such file or directory
/tmp/mambafhtXDHueer5: line 3: micromamba: command not found
/tmp/hello world/nextstrain/cli/runtimes/managed-conda/bin/.openmpi-post-link.sh: line 3: $PREFIX/.messages.txt: ambiguous redirect
/tmp/hello world/nextstrain/cli/runtimes/managed-conda/bin/.openmpi-post-link.sh: line 4: $PREFIX/.messages.txt: ambiguous redirect
/tmp/hello world/nextstrain/cli/runtimes/managed-conda/bin/.openmpi-post-link.sh: line 5: $PREFIX/.messages.txt: ambiguous redirect
/tmp/hello world/nextstrain/cli/runtimes/managed-conda/bin/.openmpi-post-link.sh: line 6: $PREFIX/.messages.txt: ambiguous redirect
/tmp/hello world/nextstrain/cli/runtimes/managed-conda/bin/.openmpi-post-link.sh: line 7: $PREFIX/.messages.txt: ambiguous redirect
/tmp/hello world/nextstrain/cli/runtimes/managed-conda/bin/.openmpi-post-link.sh: line 8: $PREFIX/.messages.txt: ambiguous redirect
/tmp/hello world/nextstrain/cli/runtimes/managed-conda/bin/.openmpi-post-link.sh: line 10: $PREFIX/.messages.txt: ambiguous redirect
/tmp/hello world/nextstrain/cli/runtimes/managed-conda/bin/.openmpi-post-link.sh: line 11: $PREFIX/.messages.txt: ambiguous redirect
/tmp/hello world/nextstrain/cli/runtimes/managed-conda/bin/.openmpi-post-link.sh: line 12: $PREFIX/.messages.txt: ambiguous redirect
/tmp/hello world/nextstrain/cli/runtimes/managed-conda/bin/.openmpi-post-link.sh: line 13: $PREFIX/.messages.txt: ambiguous redirect
/tmp/hello world/nextstrain/cli/runtimes/managed-conda/bin/.openmpi-post-link.sh: line 14: $PREFIX/.messages.txt: ambiguous redirect
/tmp/hello world/nextstrain/cli/runtimes/managed-conda/bin/.openmpi-post-link.sh: line 15: $PREFIX/.messages.txt: ambiguous redirect
/tmp/hello world/nextstrain/cli/runtimes/managed-conda/bin/.openmpi-post-link.sh: line 16: $PREFIX/.messages.txt: ambiguous redirect
/tmp/hello world/nextstrain/cli/runtimes/managed-conda/bin/.openmpi-post-link.sh: line 17: $PREFIX/.messages.txt: ambiguous redirect
```

This is sloppy packaging, but the things that fail don't seem load-bearing. We can try to address in the upstream Conda Forge feedstock recipes if we want.

Reply:
The packaging bug appears to be limited to custom post-link scripts in package recipes that don't quote `$PREFIX` (which is supplied by conda/mamba/micromamba). For example:

https://github.com/conda-forge/gdk-pixbuf-feedstock/blob/c505fd062f4ec5b9d253a0b55c7ffb48bc3bc7e6/recipe/post-link.sh#L12

We could fix that with a patch like this:

```diff
diff --git a/recipe/post-link.sh b/recipe/post-link.sh
index b4c2663..9fb7e11 100644
--- a/recipe/post-link.sh
+++ b/recipe/post-link.sh
@@ -9,7 +9,7 @@
 # available as `$PREFIX/bin/.gdk-pixbuf-post-link.sh`.
 
 set -e
-eval $(grep -v : "$PREFIX/lib/pkgconfig/gdk-pixbuf-2.0.pc")
+eval "$(grep -v : "$PREFIX/lib/pkgconfig/gdk-pixbuf-2.0.pc")"
 if [ -z "$gdk_pixbuf_cache_file" ] ; then
     bindir="$(echo "${PREFIX}/lib/gdk-pixbuf-2.0/"*)"
     if [ ! -d "$bindir" ] ; then
```

But that reveals another issue, one more fundamental to the [admittedly-hacky](https://github.com/conda-forge/gdk-pixbuf-feedstock/blob/c505fd062f4ec5b9d253a0b55c7ffb48bc3bc7e6/recipe/post-link.sh#L3-L5) technique used to source a subset of the pkgconfig file (`.pc`) as a shell script: the file isn't _actually_ shell syntax compatible, as it doesn't quote variable values (presumably because the actual `pkgconfig` command which is normally used to read those `.pc` files parses them directly).

```
$ grep -v : /tmp/hello\ world/nextstrain/cli/runtimes/managed-conda/lib/pkgconfig/gdk-pixbuf-2.0.pc
prefix=/tmp/hello world/nextstrain/cli/runtimes/managed-conda
libdir=${prefix}/lib
includedir=${prefix}/include

bindir=${prefix}/bin
…
```

Another, more innocuous example of not quoting `$PREFIX` is https://github.com/conda-forge/openmpi-feedstock/blob/5de9e8045ac72c4537b57f9a0bac3acf85bff069/recipe/post-link.sh.

Reply:
My inclination is not to worry about this now as I don't think it ultimately affects the functionality we need. But it'd be worth raising as issues on the relevant Conda Forge feedstock repos, or even submitting a PR for them if we're so inclined.

Reply:
This patch might do it for gdk-pixbuf-feedstock:

```patch
diff --git a/recipe/post-link.sh b/recipe/post-link.sh
index b4c2663..7212a98 100644
--- a/recipe/post-link.sh
+++ b/recipe/post-link.sh
@@ -9,7 +9,7 @@
 # available as `$PREFIX/bin/.gdk-pixbuf-post-link.sh`.
 
 set -e
-eval $(grep -v : "$PREFIX/lib/pkgconfig/gdk-pixbuf-2.0.pc")
+eval "$(grep -v : "$PREFIX/lib/pkgconfig/gdk-pixbuf-2.0.pc" | sed -E 's/=(.+)$/="\1"/')"
 if [ -z "$gdk_pixbuf_cache_file" ] ; then
     bindir="$(echo "${PREFIX}/lib/gdk-pixbuf-2.0/"*)"
     if [ ! -d "$bindir" ] ; then
```

but I'd have to verify and aren't sure how to build a Conda Forge package locally yet.

Reply:
Oh, hmm. There's a larger issue here that does impact our usage. The prefix gets hardcoded/embedded into the shebang lines of programs in the Conda env (a standard practice), but shebangs have no escaping/quoting mechanism so paths with spaces simply can't be used. We could potentially work around this with symlinks, or we could adjust the install prefix for macOS to not use the standard location. Need to think on this more.

Reply:
FWIW, Conda installs into `~/miniconda3/envs/auspice/bin/node` by default. We already have a `~/.nextstrain` directory so we could use that...

Reply:
Nod. I wanted to avoid using `~/.nextstrain` for non-config data if possible, as is convention. I suppose we could put a symlink in there though, or just not worry about it.

Reply:
`mamba` does output this warning:

> WARNING: A space was detected in your requested environment path
'/tmp/hello world/gdk'
Spaces in paths can sometimes be problematic.

Reply:
Apparently "Microsoft intentionally made programs install to C:\Program Files on Windows 95+ to force programmers to deal with spaces in filenames". Doesn't seem to have worked 😉

Reply:
https://github.com/conda-forge/gdk-pixbuf-feedstock/pull/34, fwiw.

Reply:
> Apparently "Microsoft intentionally made programs install to C:\Program Files on Windows 95+ to force programmers to deal with spaces in filenames". Doesn't seem to have worked wink

Well, it worked _for Windows programmers_. :-)

Reply:
> We could potentially work around this with symlinks, or we could adjust the install prefix for macOS to not use the standard location. Need to think on this more.

I [roughed out this symlink workaround](https://gist.github.com/tsibley/766a1945b505ad54e80906f1da45726c), but unfortunately it's thwarted because Micromamba (and probably Mamba/Conda too?) resolves the given `--prefix` path to the real path. Whomp.

While I'd like to DTRT and use the platform-intended locations for these files, I think I'm ready to throw in the towel and locate them under `~/.nextstrain/` with our existing config files.

Reply:
The other consideration then is where the standalone installers put `nextstrain`. Currently it's in the platform-intended user app data dirs, but if we're not using those for the managed Conda runtime, then it'd probably be better to stay consistent and always use paths under `~/.nextstrain`.

Reply:
We could use `~/.local` for macOS too – I know it's Linux convention, but `gh cli` puts a file in there anyways:

```
% cat ~/.local/state/gh/state.yml 
checked_for_update_at: 2022-08-25T22:42:17.492502-07:00
latest_release:
    version: v2.14.7
    url: https://github.com/cli/cli/releases/tag/v2.14.7
    publishedat: 2022-08-25T09:30:22Z
```

Reply:
Hmm. I think I'm still inclined to put everything under `~/.nextstrain` at this point unless we identify problems that may cause.

The typical reason for segregating app config from app data is because the config is often portable across machines while the data is not (or can be but isn't desired to be). This comes up particularly in thin client networks and/or cluster environments where parts of your home dir may be provisioned via a network file system but other parts aren't. I'm not sure this will matter much in practice though for our users.

Reply:
Repush on Friday switched to `~/.nextstrain/runtimes/managed-conda/{micromamba,env}` for the relevant paths.

Reply:
> While I'd like to DTRT and use the platform-intended locations for these files, I think I'm ready to throw in the towel and locate them under ~/.nextstrain/ with our existing config files.

I think this is the right call. Making PRs to non-conforming packages will turn into a long-term game of whack-a-mole!